### PR TITLE
[feat] OSS: Sync all attributes

### LINKED ISSUE #66: [feat] Sync OSS.param_groups and the sharded optimizer param_groups
## 🚀 Feature
Make sure that all attributes (not just LR) are in sync in between the OSS.param_groups and the actual optimizer. 
 
## Motivation
Some frameworks make it possible to alter any attribute here, not just LR (momentum, beta for Adam, ..). We do not currently support this and silently fail

## Pitch
Part of the not really well defined PyTorch optimizer features which are nice to have

## Alternatives
- At least add a warning when we're out of sync
- Nuke the .param_groups access
- Force a new OSS optimizer to be created everytime the user wants to change an attribute in param_groups

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting

- [-] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [-] Did you make sure to update the docs?
- [x] Did you write any new necessary tests? (lr scheduler test)

## What does this PR do?
Fixes #66

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by blefaudeux** [comment]
cc @mannatsingh

**Review by msbaines — CHANGES_REQUESTED** [review]
Please add a test.

Does this fix a bug? The only attribute that is common to all optimizers is learning rate.

**Comment by blefaudeux** [comment]
> Please add a test.
> 
> Does this fix a bug? The only attribute that is common to all optimizers is learning rate.

see for instance https://github.com/facebookresearch/ClassyVision/blob/master/classy_vision/optim/classy_optimizer.py#L162
in classy the user is free to pass any callback that he/she wants to adjust any setting over time, can be LR but can also be something else, the "lr" key is not specific as you can see in the code snippet. Not syncing them means silent failure for the end user when using oss, not very nice.

I'll add a non-lr unit test, good point, I was a bit lazy and thought that lr could be enough but that's wrong

**Comment by msbaines** [comment]
> > Please add a test.
> > Does this fix a bug? The only attribute that is common to all optimizers is learning rate.
> 
> see for instance https://github.com/facebookresearch/ClassyVision/blob/master/classy_vision/optim/classy_optimizer.py#L162
> in classy the user is free to pass any callback that he/she wants to adjust any setting over time, can be LR but can also be something else, the "lr" key is not specific as you can see in the code snippet. Not syncing them means silent failure for the end user when using oss, not very nice.
> 
> I'll add a non-lr unit test, good point, I was a bit lazy and thought that lr could be enough but that's wrong

It would be nice to capture some of this discussion in the commit message or via a code comment.

**Comment by blefaudeux** [comment]
> > > Please add a test.
> > > Does this fix a bug? The only attribute that is common to all optimizers is learning rate.
> > 
> > 
> > see for instance https://github.com/facebookresearch/ClassyVision/blob/master/classy_vision/optim/classy_optimizer.py#L162
> > in classy the user is free to pass any callback that he/she wants to adjust any setting over time, can be LR but can also be something else, the "lr" key is not specific as you can see in the code snippet. Not syncing them means silent failure for the end user when using oss, not very nice.
> > I'll add a non-lr unit test, good point, I was a bit lazy and thought that lr could be enough but that's wrong
> 
> It would be nice to capture some of this discussion in the commit message or via a code comment.

ah, the intent was to have this documented in the linked issue (https://github.com/facebookresearch/fairscale/issues/66), but if it's easier I can elaborate more in the commit messages or PR ? Alrighty, learning by doing, there are some hidden assumptions

**Comment by msbaines** [comment]
> > > > Please add a test.
> > > > Does this fix a bug? The only attribute that is common to all optimizers is learning rate.
> > > 
> > > 
> > > see for instance https://github.com/facebookresearch/ClassyVision/blob/master/classy_vision/optim/classy_optimizer.py#L162
> > > in classy the user is free to pass any callback that he/she wants to adjust any setting over time, can be LR but can also be something else, the "lr" key is not specific as you can see in the code snippet. Not syncing them means silent failure for the end user when using oss, not very nice.
> > > I'll add a non-lr unit test, good point, I was a bit lazy and thought that lr could be enough but that's wrong
> > 
> > 
> > It would be nice to capture some of this discussion in the commit message or via a code comment.
> 
> ah, the intent was to have this documented in the linked issue (#66), but if it's easier I can elaborate more in the commit messages or PR ? Alrighty, learning by doing, there are some hidden assumptions

The linked issue is two levels of indirection from the code. It's nice to have a useful summary in the commit message. The commit message is one level of indirection away from the code. When maintaining code, it is common to `git log` or `git blame` a file. Understanding the rationale for a change from the code or the commit message is very handy. You have all the information available in the log. Reading the issue requires you to move your attention away from the log.

https://chris.beams.io/posts/git-commit/#why-not-how

**Review by msbaines — APPROVED** [review]
Please add more context to the commit message before landing.
